src/sources/docker_logs.rs

```diff
@@ -1346,11 +1360,48 @@ mod integration_tests {
         );
     }
 
+    #[tokio::test]
+    async fn exclude_containers() {
+        trace_init();
+
+        let will_be_read = "12";
+        let all_containers = "vector_test";
+        let excluded0 = "vector_test_exclude_0";
+        let included0 = "vector_test_include_0";
+        let included1 = "vector_test_include_1";
+
+        let docker = docker().unwrap();
+
+        let out = source_with_config(DockerLogsConfig {
+            include_containers: Some(vec![all_containers.to_owned()]),
+            exclude_containers: Some(vec![excluded0.to_owned()]),
+            ..DockerLogsConfig::default()
+        });
+
+        let id0 = container_log_n(1, excluded0, None, "will not be read", &docker).await;
+        let id1 = container_log_n(1, included0, None, will_be_read, &docker).await;
+        let id2 = container_log_n(1, included1, None, will_be_read, &docker).await;
+        let events = collect_n(out, 2).await.unwrap();
+        container_remove(&id0, &docker).await;
+        container_remove(&id1, &docker).await;
+        container_remove(&id2, &docker).await;
+
+        assert_eq!(
```

Comment:
I'm actually wondering if this wouldn't pass sometimes, even if the excluded container was included, since we `collect_n(out, 2)` and, I believe?, is no guaranteeing of the ordering of the logs (that is to say, that `will not be read` might be the third log).

This is probably an edge case, but we might be able to cover it by shutting down the source and then using `collect_ready()` to collect _all_ events.

I'm curious what @bruceg thinks.

Reply:
Good call. I've switched this out for `collect_ready` and added an additional check that only 2 events are collected.